Enabled nullable reference types on Avalonia.Markup

jp2masa opened the pull request:
## Changes

Enabled nullable reference types on `Avalonia.Markup`. Tried to change as little as possible. It's still disabled on a few lines in `Binding.cs`, as I didn't understand exactly what the code should do if `(target as IStyledElement) ?? (anchor as IStyledElement)` is `null`, and it's really unlikely to be.

jp2masa committed:
Enabled nullable reference types on Avalonia.Markup.

Gillibald committed:
Merge branch 'master' into nullable

maxkatz6 commented:
> I didn't understand exactly what the code should do if (target as IStyledElement) ?? (anchor as IStyledElement) is null, and it's really unlikely to be.

If it will be "null" app will fail on Contract.Requires check inside of CreateExpressionObserver method. I believe it's safe to throw an exception there.

grokys commented on src/Markup/Avalonia.Markup/Data/Binding.cs:
```diff
@@ -59,20 +59,21 @@ public Binding(string path, BindingMode mode = BindingMode.Default)
+        protected override ExpressionObserver CreateExpressionObserver(IAvaloniaObject target, AvaloniaProperty targetProperty, object? anchor, bool enableDataValidation)
         {
-            Contract.Requires<ArgumentNullException>(target != null);
+            Contract.Requires<ArgumentNullException>(target is not null);
```
Note that when I add nullable reference types, I remove `Contract.Requires` and replace it with `arg = arg ?? throw new ArgumentNullException(nameof(arg))` because:

- It's nearly as succinct
- It plays nicer with NRTs
- It actually includes the name of the arg in the error message

jp2masa replied:
I tried to keep changes minimal, but I can change that (also because `is not` is a C# 9 feature and it breaks the build).

grokys commented on src/Markup/Avalonia.Markup/Markup/Parsers/Nodes/StringIndexerNode.cs:
```diff
@@ -29,15 +29,15 @@ protected override bool SetTargetValueCore(object value, BindingPriority priorit
             var dictionary = target as IDictionary;
             var indexerProperty = GetIndexer(typeInfo);
-            var indexerParameters = indexerProperty?.GetIndexParameters();
+            ParameterInfo[] indexerParameters;
```
Why move this initialization into the `if`? It's slightly less readable like this to me, though I don't really mind either way, just wondering what the reason is.

jp2masa replied:
IIRC, it had a warning because `indexerParameters` could be `null` (in `indexerParameters.Length == Arguments.Count`), although we know it's not, because it's only `null` when `indexerProperty` is `null` as well, but the compiler doesn't know that. I also agree that it's slightly less readable.

jp2masa commented:
@maxkatz6 Throw an exception where? Something like this?

```cs
(target as IStyledElement) ?? (anchor as IStyledElement) ?? throw new InvalidOperationException();
```

Or change the corresponding parameter to nullable in `CreateElementObserver`, `CreateSourceObserver`, `CreateTemplatedParentObserver` and `CreateFindAncestorObserver`, and keep the `Contract` throwing? I think this wouldn't make much sense as the contract should match what the parameter asks for.

maxkatz6 commented:
@jp2masa yes, check for null before passing it as an argument.
Or use "!" operator with some comment because it will unlikely be null, and we already have null-check in the methods to which we pass this parameter.

grokys committed:
Merge branch 'master' into nullable

grokys committed:
Remove Contract usages.

grokys committed:
Fix nullable issues in Binding.cs,

grokys commented:
@jp2masa fixing the remaining issues myself on your branch, hope that's OK.

grokys committed:
Fix nullable annotation.

grokys committed:
Add #nullable enable on shared files.

Some files in Avalonia.Markup are linked into other projects and their use of `?` was causing warnings in the other projects. Explicitly add `#nullable enable` in these cases for now.

grokys committed:
Fix nullability of Equals methods.

grokys committed:
Promote nullable warnings to errors.

grokys committed:
Merge branch 'master' into nullable
